docs/specifications/tests/Basic-TP/basic01.md

```diff
@@ -1,199 +1,311 @@
+B01_NO_CHILD               |ERROR  | domain_child, domain_super            | "{domain_child}" does not exist as a DNS zone. Try to test "{domain_super}" instead.
+B01_PARENT_FOUND           |INFO   | domain, ns_ip_list                    | The parent zone is "{domain}" as returned from name servers "{ns_ip_list}".
+B01_PARENT_INDETERMINED    |WARNING| ns_ip_list                            | The parent zone cannot be determined on name servers "{ns_ip_list}".
+B01_UNEXPECTED_NS_RESPONSE |WARNING|domain_child, domain_parent, ns_ip_list| Name servers for parent domain "{domain_parent}" give an incorrect response on SOA query for "{domain_child}". Returned from name servers {ns_ip_list}.
+
+
+The value in the Level column is the default severity level of the message. The
+severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+[Severity Level Definitions] document.
+
+The argument names in the Arguments column lists the arguments used in the
+message. The argument names are defined in the [argument list].
+
+
+## Test procedure
+
+In this section and unless otherwise specified below, the terms "[DNS Query]"
+follow the specification for DNS queries as specified in
+[DNS Query and Response Defaults]. The handling of the DNS responses on the
+DNS queries follow, unless otherwise specified below, what is specified for
+[DNS Response] in the same specification.
+
+1. If the *Child Zone* is the root zone (".") then:
+   1. Output *[B01_CHILD_FOUND]* with zone name (".").
+   2. Exit the test case.
+
+2. Create [DNS queries][DNS Query]:
+   1. Query type SOA and query name *Child Zone* ("SOA Child Query").
+   2. Query type DNAME and query name *Child Zone* ("DNAME Query").
+
+3. Create the following empty sets:
+   1. Parent name server IP, parent zone name and DNS response
+      ("Parent Name Server IP").
+   2. Parent name server IP and parent zone name ("Parent Found").
+   3. Parent name server IP and parent zone name ("Delegation Found").
+   4. Parent name server IP and parent zone name ("Non-AA Non-Delegation Found")
+   5. Parent name server IP and parent zone name ("AA NXDomain Found").
+   6. Parent name server IP and parent zone name ("AA SOA Found").
+   7. Parent name server IP and parent zone name ("AA CNAME Found").
+   8. Parent name server IP, parent zone name and DNAME target
+      ("AA DNAME Found").
+   9. Parent name server IP and parent zone name ("AA NODATA Found").
+
+4. Find the parent zone of the *Child Zone* by iteratively
+   [sending](#terminology) *SOA Child Query* to all name servers found. Start by
+   using the nameservers from the *Root Name Servers*.
+   1. Follow all paths from root and downwards by following the referrals
+      (non-AA response with empty answer section and NS records in the authority
+      section).
```

Comment:
There's a pretty standard way to perform searches like this. Have you considered doing using this approach?

```
1. Create empty sets:
   1. Name server IP ("Remaining Server Addresses").
   2. Name server IP ("Handled Server Addresses").
...
4. Insert all addresses from *Root Name Servers* into *Remaining Server Addresses*.
5. While *Remaining Server Addresses* is non-empty:
   1. Remove an address ("Server Address") from *Remaining Server Addresses*.
   3. Insert *Server Address* into *Handled Server Addresses*.
   4. Send query to *Server* and collect interesting addresses ("Candidates") from the response.
   5. Remove any addresses from *Candidates* that are also present in *Handled Server Addresses*.
   6. Insert all addresses from *Candidates* into *Remaining Server Addresses*.
```

This way it's unambiguous what it means to follow all paths.

Reply:
Updated to something like it.